Updates CoreFile model to support many-to-many associations with both…

candyhazlett opened the pull request:
Resolves issue #110 by:

- implementing proper many-to-many associations with Collections and Projects via join tables
- adding validation and helper methods for managing CoreFile relationships
- updating views logic to handle new association structure and display correct associated core file count
- resolving Rails 5.2 ActiveStorage threading deadlock by switching ActiveJob from async adapter to  background processing after discovering a known [error](https://github.com/rails/rails/issues/34939) 
- configures Resque initialization to handle ActiveStorage queue
- updates dummy_data_generator rake tasks to work with new associations and process background jobs

Steps to test
Set up the database:
- [ ] Delete existing core files, collections, and projects:`rails runner "CoreFile.destroy_all; Collection.destroy_all; Project.destroy_all"`
- [ ] Then regenerate from terminal: `rails dummy_data_generator:non_user_records`
- [ ] In a separate terminal window, restart Resque: `QUEUE=* bundle exec rake resque:work`
- [ ] Start the server in another terminal window: `rails server`

Verify Core File Associations in Browser:
  - Visit http://localhost:3000
  - Navigate to Projects browse
  - Click on a project row dropdown arrow and observe CoreFiles listed
  - Navigate to Collections (from Project) and verify CoreFiles appear there too

Verify Many-to-Many Associations in Rails Console:
- [ ] Open rails console from terminal: `rails console`
- [ ] Test that CoreFiles can belong to multiple Collections: `cf = CoreFile.first
  cf.collections.count`
- [ ] Test that CoreFiles can't belong to multiple Projects: `cf.projects.count`
- [ ] Test adding/removing associations: `collection = Collection.last
  cf.collections << collection
  cf.collections.count`
- [ ] Verify bidirectional association: `collection.core_files.include?(cf)`
- [ ] Test Background Job Processing: `RebuildReadingInterfaceJob.perform_later(cf.id)`
- [ ] Check Resque to confirm job count: `Resque.info`

candyhazlett committed:
Updates CoreFile model to support many-to-many associations with both Collections and Projects. Adds proper association helpers, validations, and UI updates for managing CoreFile relationships. Resolves issue of rake task for generating dummy CoreFiles failing silently due to Rails 5.2 ActiveStorage threading deadlock. Fixed by switching ActiveJob from async adapter to Resque for background job processing.

amclark42 commented on app/models/image_file.rb:
```diff
@@ -15,6 +15,17 @@ def process_image_data
-    errors.add(:image, "must be a JPEG, JPG, GIF, or PNG file") unless content_type.in?(%w[image/jpeg image/jpg image/gif image/png])
+    valid_types = %w[
+      image/jpeg
+      image/jpg
```
[Only `image/jpeg` is a valid MIME type.](https://stackoverflow.com/questions/33692835/is-the-mime-type-image-jpg-the-same-as-image-jpeg#37266399) Let's remove `image/jpg`.

candyhazlett replied:
I meant to look up which is valid. Thanks for catching this!

candyhazlett replied:
Resolved w/commit cb66e0c

amclark42 commented on app/views/core_files/_data_form.html.haml:
```diff
@@ -61,11 +61,11 @@
               = f.label :remove_thumbnail, "Remove Image File"
               = f.check_box "remove_thumbnail", {:checked => false}
-            = f.file_field :thumbnails, accept: ".jpeg, .jpg, .png", direct_upload: true, multiple: true
+            = f.file_field :thumbnails, accept: ".jpeg, .jpg, .png, .gif, .svg, .pdf", direct_upload: true, multiple: true
```
It'd be useful to make sure that the accepted MIME types in `models/image_files.rb` are kept in sync with this input field. We could use a hash (not sure if that's the right terminology for Ruby) in the ImageFile model to associate a MIME type with the extensions, and populate this accept attribute with both the extensions and the MIME types.

amclark42 replied:
There's a similar field in the Projects data form. I'd like to see the thumbnails/images fields separated out into a template that both model types could use (and collections too, when we get there).

candyhazlett replied:
'Hash' is the right term and the image file model does validate file format currently:
 ```
 validate :file_format

  ...

  def file_format
    return unless file.attached?

    valid_types = %w[
      image/jpeg
      image/gif
      image/png
      image/svg+xml
      application/pdf
    ]

    unless content_type.in?(valid_types)
      errors.add(:image, "must be a JPEG, GIF, PNG, SVG, or PDF file")
    end
```

We can remove the client-side validation so there's just one place to keep track of.

candyhazlett replied:
Resolved w/commit cb66e0c

amclark42 commented on lib/tasks/dummy_data_generator.rake:
```diff
@@ -3,6 +3,35 @@ namespace :dummy_data_generator do
   require 'open-uri'
 
+  # Helper method to check if Resque workers are running
+  def check_resque_workers
```
It might be useful to have another method to check if Solr's running. My first attempt testing this PR crashed because I'd forgotten to fire up Solr. >_>;;

candyhazlett replied:
Resolved w/commit cb66e0c

amclark42 commented:
@candyhazlett As I'm going through the PR diff, I'm seeing a lot of changes to the image file code. Could you write up a bit of context for that? How should I test the changes?

amclark42 commented on lib/tasks/dummy_data_generator.rake:
```diff
@@ -126,39 +155,116 @@ namespace :dummy_data_generator do
+            ActiveRecord::Base.connection_pool.release_connection
+            sleep 0.3
+            puts "  -> Cleared connection pool after #{i + 1} files"
+          end
```
I believe this is the workaround for the known error? Looks good. I recommend adding a longer comment that describes this as a workaround, and gives a hint as to when we'd want to remove it (probably when we upgrade past Rails v5.2?).

candyhazlett replied:
No, that's left over from my many (ultimately) failed attempts to resolve the threading lock up issue.  I'll remove it since active_job handles this now.

candyhazlett replied:
Resolved w/commit cb66e0c

amclark42 commented on lib/tasks/dummy_data_generator.rake:
```diff
@@ -126,39 +155,116 @@ namespace :dummy_data_generator do
 
   desc 'generate all dummy data'
   task :run_all => :environment do
+    # Check for Resque workers before proceeding
+    check_resque_workers
```
It looks like `check_resque_workers` is only called at the beginning of `run_all` and `core_files`. Is Resque just required for creating Core Files, or is it necessary for basically anything that creates a record?

candyhazlett replied:
Good question! I ran into the issue during core file creation. But since collections, users, and projects can have attached files as well, the workers should be started when the app starts in Production. 

For the rake tasks, if the create project, collections, and/or users tasks are updated to also attach an image file to the newly created record, then yes, those tasks should also call check_resque_workers.

candyhazlett replied:
Resolved w/commit cb66e0c

amclark42 reviewed (changes requested):
### Process

#### Setting up the database

I ran into some trouble while following the "Steps to test" as written — the last 13 projects created had no collections, core files, or regular project contributors (just a depositor). I was able to get the expected numbers after running these commands:

```shell
rails db:drop
rails db:create
rails db:migrate
rails dummy_data_generator:run_all
```

With these commands, I was able to see in the browser that every project had 3 collections and 15 documents.

I'm not sure how useful it is to investigate whether there was an issue with the `destroy_all` commands, or the `dummy_data_generator:non_user_records` task, or something in my outdated databases. Happy to run further tests on request.

#### Testing core files across multiple collections

Following the instructions for testing in the Rails console, I found that core files can indeed be associated with multiple collections, and the relationship goes both ways.

However, I want to check in on "Test that CoreFiles can belong to multiple Projects". A Core File should only ever be associated with a single Project at a time. (Issue #110: "core files' within multiple collections maintain the same parent project".)

By taking the first Core File and the last Collection, the testing steps ensure that the core file and collection will be in different projects before I manually add the core file to the collection. This update shouldn't be allowed, but it occurs without complaint.

When I ran `cf.projects.count` again, it was still at `1` in spite of the multiple projects. I suspect that means something is still messed up with the association between projects and core files. The generator script is doing the right thing, but manual updates may not.

```
2.7.7 :001 > cf = CoreFile.first
 => #<CoreFile id: 1, title: "Sleep the Brave", description: "Crime/Detective", created_at: "2025-11-06 19:54:29", upda... 
2.7.7 :002 > cf.collections.count
 => 1 
2.7.7 :003 > cf.projects.count
 => 1 
2.7.7 :004 > collection = Collection.last
 => #<Collection id: 75, title: "Scotch Eggs", description: "A few vices are sufficient to darken many virtues.", creat... 
2.7.7 :005 > cf.collections << collection
 => #<ActiveRecord::Associations::CollectionProxy [#<Collection id: 1, title: "Ricotta Stuffed Ravioli", description: "A few vices are sufficient to darken many virtues.", created_at: "2025-11-06 19:54:23", updated_at: "2025-11-06 19:54:23", project_id: 1, is_public: true, depositor_id: 36, discarded_at: nil>, #<Collection id: 75, title: "Scotch Eggs", description: "A few vices are sufficient to darken many virtues.", created_at: "2025-11-06 19:54:29", updated_at: "2025-11-06 19:54:29", project_id: 25, is_public: false, depositor_id: 255, discarded_at: nil>]> 
2.7.7 :006 > cf.collections.count
 => 2 
2.7.7 :007 > cf.projects.count
 => 1 
2.7.7 :008 > collection.core_files.include?(cf)
 => true 
```

^ Note that `cf` appears in collections associated with project 1 and project 25. That's a problem.

#### Initiating the reading interface job

I wasn't able to run `RebuildReadingInterfaceJob.perform_later()` (which doesn't exist) or `RebuildReadingInterfaceJob.perform()` (which does). Do you have uncommitted changes in that file?

```
2.7.7 :009 > RebuildReadingInterfaceJob.perform_later(cf.id)
Traceback (most recent call last):
        1: from (irb):9
NoMethodError (undefined method `perform_later' for RebuildReadingInterfaceJob:Class)
Did you mean?  perform
2.7.7 :010 > RebuildReadingInterfaceJob.perform(cf.id)
Traceback (most recent call last):
        4: from (irb):9
        3: from (irb):10:in `rescue in irb_binding'
        2: from app/jobs/rebuild_reading_interface_job.rb:33:in `perform'
        1: from app/jobs/rebuild_reading_interface_job.rb:37:in `rescue in perform'
NoMethodError (undefined method `errors_display=' for #<CoreFile:0x000000010d4ea8f0>)
```

#### Updates to the image files

The PR description doesn't mention the changes to the Image File model, and downstream changes to various views and test files. I'd like to have a better sense of what these changes are and how I should test them before approving this PR.

### Review

This PR does fix my ability to see the number of core files, and includes a workaround for the known bug in our current version of ActiveStorage. Unfortunately, I think the fix needs a little more TLC.

I'd like to see these issues addressed before I approve the PR:

1. Ensure that core files and projects are correctly associated, even on manual creation or update.
2. Provide context and test procedure(s) for the Image File changes.
3. Remove `image/jpg` from acceptable MIME types for images.
4. Decide whether to update in this PR, or create new issues for:
    1. `RebuildReadingInterfaceJob.perform_later()`
    2. Enforce that Core Files should only belong to 1 Project, and can't be placed in Collections across multiple Projects.

I'd also love to hear your read on my problems running the test procedures — is this something we should spend more time debugging?

candyhazlett committed:
Remove jpg from list of acceptable file types on project, collection, and core file data form, as well as the avatar_uploader, and collection, project, and zip content validators. Remove unused delay logic in core file creation task. This has been replaced with resque workers.

candyhazlett commented on app/models/image_file.rb:
`valid_types` array added to `file_format` validation method; each supported image file type included as a string: `type/subtype` to better adhere to Rails conventions.

candyhazlett commented:
@amclark42, So, the testing step for the core_file-project association should have said, "Test that core files _can't_ belong to multiple projects." I've fixed that typo to accurately represent the expected behavior. 

And thanks very much for catching that single core_file with multiple projects bug! That was huge. For some goofy reason, `has_and_belongs_to_many` associations in Rails will completely bypass validations and write the invalid record to the db when an object is modified directly, e.g., `<<`, `.delete`, etc,  rather than`.save` or `.valid?`, which will trigger the validation. 

This was kind of a bear to debug because when first I tested it the way you did, I got this: 
```
cf = CoreFile.first
=> #<CoreFile id: 1, title: "Now Sleeps the Crimson Petal", description: "Western...
irb(main):002:0> cf.collections.count
=> 2
irb(main):003:0> cf.projects.count
=> 1
irb(main):004:0> collection = Collection.last
=> #<Collection id: 75, title: "Cheeseburger", description: "Good habits formed a...
irb(main):005:0> cf.collections << collection
Traceback (most recent call last):
        1: from (irb):5
ActiveRecord::RecordNotUnique (Mysql2::Error: Duplicate entry '75-1' for key 'collections_core_files.index_collections_core_files_on_collection_id_and_core_file_id': INSERT INTO `collections_core_files` (`collection_id`, `core_file_id`) VALUES (75, 1))
irb(main):006:0> cf.collections.count
=> 2
irb(main):007:0> cf.projects.count
=> 1
irb(main):008:0> collection.core_files.include?(cf)
=> true
irb(main):009:0> cf.projects.map(&:id)
=> [1]
```

But then I noticed the core file was in two collections already, so I tried again with a core file that was only in one. That's when I saw what you saw:
```
irb(main):010:0> cf = CoreFile.all.select { |c| c.collections.count.eql? 1 }.first
=> #<CoreFile id: 2, title: "If Not Now, When?", description: "Western", created_at: "2025-11-07 20:43:00...
irb(main):011:0> cf.collections.count
=> 1
irb(main):012:0> cf.projects.count
=> 1
irb(main):013:0> collection = Collection.last
=> #<Collection id: 75, title: "Cheeseburger", description: "Good habits formed at youth make all the dif...
irb(main):014:0> cf.collections.map(&:id)
=> [1]
irb(main):015:0> cf.projects.map(&:id)
=> [1]
irb(main):016:0> collection.id
=> 75
irb(main):017:0>  cf.collections << collection
=> #<ActiveRecord::Associations::CollectionProxy [#<Collection id: 1, title: "Pasta Carbonara", description: "Control thy passions lest they take vengence on th...", created_at: "2025-11-07 20:42:49", updated_at: "2025-11-07 20:42:49", project_id: 1, is_public: true, depositor_id: 108, discarded_at: nil>, #<Collection id: 75, title: "Cheeseburger", description: "Good habits formed at youth make all the differenc...", created_at: "2025-11-07 20:43:00", updated_at: "2025-11-07 20:43:00", project_id: 25, is_public: false, depositor_id: 249, discarded_at: nil>]>
```

The solution is to go back to the `has_many_through` association I removed [here](https://github.com/NEU-DSG/tapas_rails/commit/9cbb04d308a85c6ab321cc1737871f61587c1b19) and add the same model validations to the join tables too so that they can't be overrode. I will make a commit on this branch once that's ready.

candyhazlett commented:
@amclark42, let's make a new issue for the `RebuildReadingInterfaceJob.perform_later()` missing method error. And let's discuss the changes to the image file and other files that aren't clearly outlined in the PR message at our next meeting.

amclark42 commented:
@candyhazlett Wow, impressive debugging! Excellent work!

Yes, let's discuss the other issues in our meeting.

candyhazlett commented on lib/tasks/dummy_data_generator.rake:
resume walk through of image file-related modifications included in this PR

candyhazlett committed:
Converts habtm associations to has_many :through pattern using CollectionsCoreFile and CoreFilesProject join models. Adds validation to ensure that a core file can only belong to collections within a single project.

candyhazlett committed:
Reverts commit cb66e0c to restore jpg as valid file extension for image file upload.

amclark42 reviewed (approved):
I dropped my existing tables and recreated them, then checked the interface and ran through a modified version of the `rails console` test sequence. It works!

```
2.7.7 :001 > cf = CoreFile.first
 => #<CoreFile id: 1, title: "Many Waters", description: "Mythopoeia", created_at: "2025-11-14 16:32:03", updated_at: "... 
2.7.7 :002 > cf.collections.count
 => 1 
2.7.7 :003 > cf.projects.count
 => 1 
2.7.7 :004 > collection = Collection.last
 => #<Collection id: 75, title: "Pasta with Tomato and Basil", description: "Know how to listen, and you will profit ev... 
2.7.7 :005 > cf.collections << collection
Traceback (most recent call last):
        1: from (irb):5
ActiveRecord::RecordInvalid (Validation failed: Core file cannot belong to collections from multiple projects. Existing project(s): 1, attempted to add project: 25)
2.7.7 :006 > # Yay!
 => nil 
2.7.7 :007 > cf.collections.count
 => 1 
```

To make sure that it was still possible to add a collection from the same project, I did this:

```
2.7.7 :018 > collection = Collection.find(3)
 => #<Collection id: 3, title: "Massaman Curry", description: "The unexamined life is not worth living.", created_at: "... 
2.7.7 :019 > cf.collections << collection
 => #<ActiveRecord::Associations::CollectionProxy [#<Collection id: 1, title: "Bruschette with Tomato", description: "The unexamined life is not worth living.", created_at: "2025-11-14 16:31:57", updated_at: "2025-11-14 16:31:57", project_id: 1, is_public: true, depositor_id: 241, discarded_at: nil>, #<Collection id: 3, title: "Massaman Curry", description: "The unexamined life is not worth living.", created_at: "2025-11-14 16:31:57", updated_at: "2025-11-14 16:31:57", project_id: 1, is_public: false, depositor_id: 102, discarded_at: nil>]> 
2.7.7 :020 > cf.collections.count
 => 2 
2.7.7 :021 > cf.projects.count
 => 1 
2.7.7 :022 > collection.core_files.include?(cf)
 => true 
```

This is excellent work! I'm happy to approve the PR.

For our future selves, we are punting on:

- The `RebuildReadingInterfaceJob` described in the test section of the PR description — that code isn't represented here.
- Completion of the Image File model changes — there is work still to do, but it is outside the scope of this PR.
    - Candace and I went over the impetus for the existing code in our [2025-11-13 working meeting](https://docs.google.com/document/d/1t7FhzDMQdRp2yYG2wC1zcwNlHhIfBbWWLWEFBnygkL8/edit?tab=t.0#heading=h.7bpu88pd0dn2). The changes are a refactoring of old code for validating an uploaded file.
    - We will create an issue describing the work still to be done.

candyhazlett committed:
Pins gem versions in use currently to fix deprecation warnings and fixes Solr indexing for edit access pserson. Adds ImageFile constant definitions (VALID_FILE_TYPES, VALID_FILE_EXTENSIONS) and updates file_format validation to conditionally allow PDFs only for CoreFile objects while fixing bug using file.content_type instead of content_type. Also improves dummy data generator with variable record counts and re-enables image attachments

candyhazlett committed:
Revert "Pins gem versions in use currently to fix deprecation warnings and fixes Solr indexing for edit access pserson. Adds ImageFile constant definitions (VALID_FILE_TYPES, VALID_FILE_EXTENSIONS) and updates file_format validation to conditionally allow PDFs only for CoreFile objects while fixing bug using file.content_type instead of content_type. Also improves dummy data generator with variable record counts and re-enables image attachments"

This reverts commit 853b6f24e3d0702af3c2c1eb4f1d73e2ec90edc8.
